### ISSUE TITLE
Failing to bind shouldn't panic

### ISSUE THREAD

**Issue by udoprog** [seanmonstar/warp#237]
Today on Windows, if you try to bind to an address which is already used, warp panics with:

```
error binding to 0.0.0.0:12345: error creating server listener: Only one usage of each socket address (protocol/network address/port) is normally permitted. (os error 10048)
```

This is due to this statement:
https://github.com/seanmonstar/warp/blob/master/src/server.rs#L64

This is an issue with programs using `#![windows_subsystem = "windows"]` since they typically don't have a terminal where the panic is printed and need some other way to provide feedback to the user.

**Comment by seanmonstar** [comment]
Hm, is there something else that a program could do in this error situation? I chose to panic because in *most* cases, there's not much a user can do but check their system for what else is already running on that port.

I'd be fine with adding an extra start method that doesn't panic.

**Comment by udoprog** [comment]
> Hm, is there something else that a program could do in this error situation?

The typical case would be that you'd like to retry with a timeout if a bind fails. If the server is a component of a larger application it might not be desirable to have that application handle the panic. Like a desktop application using OAuth 2.0 by temporarily setting up a local server.

I'm using warp in a system where the server is a component that can be reloaded. I sporadically encounter issues when binding because the OS takes a while to free the port for whatever reason (even with `SO_REUSEADDR`).

I can look into adding a `try_bind`. Would you prefer if this resolves into a fallible `Future`, or return `Result<impl Future<...>, Error>`?

**Comment by seanmonstar** [comment]
I kinda find `Result<Future>`s a little odd... I guess they make sense, but it reminds me of like in JavaScript, where an function *might* throw immediately, or include the error in the `Promise`, and how I so wish it'd all just be in the `Promise`...

**Comment by seanmonstar** [comment]
New methods added in #250, published in v0.1.17.
